fuel-core-interfaces/src/relayer.rs

```diff
@@ -0,0 +1,143 @@
+pub struct DepositCoin {
+    pub owner: Address,
+    pub amount: Word,
+    pub asset_id: AssetId,
+    pub eth_block_deposited: u64,
+    pub fuel_block_spend: u64,
+}
+
+// Database has two main functionalities, ValidatorSet and TokenDeposits.
+// From relayer perspective TokenDeposits are just insert when they get finalized.
+// But for ValidatorSet, It is litle bit different.
+#[async_trait]
+pub trait RelayerDb:
+     Storage<Bytes32, DepositCoin, Error = KvStoreError> // token deposit
+    + Storage<Address, u64,Error = KvStoreError> // validator set
+    + Storage<u64, HashMap<Address, u64>,Error = KvStoreError> //validator set diff
+    + Send
+    + Sync
+{
+    /// deposit token to database. Token deposits are not revertable
+    async fn insert_token_deposit(
+        &mut self,
+        deposit_nonce: Bytes32, // this is ID
+        eth_block_deposited: u64, // eth block when deposit is made
+        owner: Address,       // owner
+        asset_id: AssetId,
+        amount: Word,
+    ) {
+        let coin = DepositCoin {
+            owner,
+            amount,
+            asset_id,
+            eth_block_deposited,
+            fuel_block_spend: 0,
+        };
+        // TODO check what id are we going to use
+        // depends on https://github.com/FuelLabs/fuel-specs/issues/106
+        let _ = Storage::<Bytes32, DepositCoin>::insert(self,&deposit_nonce,&coin);
+    }
+
+    /// Asumption is that validator state is already checked in eth side, and we can blidly apply
+    /// changed to db without checking if we have enought stake to reduce.
+    /// What needs to be done is to have validator set state and diff as saparate database values.
+    async fn insert_validator_set_diff(&mut self, eth_height: u64, stakes: &HashMap<Address, u64>) {
+        let _ = Storage::<u64,HashMap<Address,u64>>::insert(self, &eth_height,stakes);
+    }
+
+    async fn apply_current_validator_set(&mut self, changes: HashMap<Address,u64>) {
+        for (ref address,ref stake) in changes {
+            let _ = Storage::<Address,u64>::insert(self,address,stake);
+        }
+    }
+
+    /// get validator set for current eth height
+    async fn current_validator_set(&self) -> HashMap<Address,u64>;
+
+    /// set last finalized fuel block. In usual case this will be
+    async fn set_current_validator_set_eth_height(&self, block: u64);
+    /// Assume it is allways set as initialization of database.
+    async fn get_current_validator_set_eth_height(&self) -> u64;
+
+    /// get stakes difference between fuel blocks. Return vector of changed (some blocks are not going to have any change)
+    async fn get_validator_set_diff(
+            &self,
+            from_fuel_block: u64,
+            to_fuel_block: Option<u64>,
+    ) -> Vec<(u64,HashMap<Address, u64>)>;
+
+    /// current best block number
+    async fn get_block_height(&self) -> u64;
+
+
+    /// set newest finalized eth block
+    async fn set_eth_finalized_block(&self, block: u64);
```

Comment:
we should rename usages of eth to da, since we may target other evm compatible chains like celestia.

Reply:
Yea this is not nice naming, i refactored it